### PR TITLE
Refactor material source code, and add error checks and tests

### PR DESCRIPTION
**Pull request by smpark7**
`GenericMoltresMaterial` and `MoltresJsonMaterial` load and interpolate group constant data provided at discrete temperature values for the use in Moltres reactor simulations. For example, if the user provides cross section data at 600K and 900K and Moltres requires data at 711K, the `Material` objects interpolate the provided cross section data to obtain the estimate at 711K.

This PR introduces the following changes:
- Refactor parent class `NuclearMaterial` and its child classes `GenericMoltresMaterial` and `MoltresJsonMaterial` to reduce code duplication.
  - Combine `dummyConstruct`, `splineConstruct`, `linearConstruct`, and `monotoneCubicConstruct` functions into a single `Construct` function in `GenericMoltresMaterial`.
  - Standardize the use of inherited `_xsec_map` and `_xsec_names` variables in `GenericMoltresMaterial` and `MoltresJsonMaterial` for common interpolation schemes (`NONE`, `LINEAR`, `SPLINE`, `MONOTONE_CUBIC`).
  - Bug fix: Don't apply linear interpolation in `MoltresJsonMaterial` if user sets `interp_type=none`.
- Add error checks to `GenericMoltresMaterial` and `MoltresJsonMaterial` to check whether the number of group constant data values matches the `num_groups` and `num_precursor_groups` parameters provided by the user.
- Add tests for `GenericMoltresMaterial` and `MoltresJsonMaterial` that test whether the material objects initialize correctly, or provide the intended error message otherwise.
- Remove unnecessary Precursor modeling in DiracHX test to avoid running into the new materials error checks.

***Note: The interpolation types (`interp_type`) are as follows:
- `NONE`: No interpolation
- `LINEAR`: Linear spline interpolation
- `SPLINE`: Cubic spline interpolation
- `MONOTONE_CUBIC`: Monotone cubic interpolation

### CODE REVIEW

**Review by samgdotson — CHANGES_REQUESTED** [review]
This looks much cleaner, good job @smpark7! I’m wondering if some of these binary files are actually built files or pre-compiled? If they’re built files they should be removed, if they’re pre compiled then that’s fine. Happy to merge once this is clarified (unless you want others to review as well).

**Comment by smpark7** [comment]
Thanks for the quick review @samgdotson! The binary files are all Exodus (.e) output files in `gold` directories. They're reference results that some of the new tests compare against. So I think they're what you consider pre-compiled files.

Let's leave it unmerged until tomorrow to give others a chance to review. I'll ping you if necessary.

**Review by LukeSeifert — APPROVED** [review]
This looks good, I just had two small things I noticed.

File: src/materials/GenericMoltresMaterial.C [hunk 1]
```diff
@@ -121,228 +104,138 @@ GenericMoltresMaterial::GenericMoltresMaterial(const InputParameters & parameter
...
-      {
         temperature.push_back(value);
         for (decltype(o) k = 0; k < o; ++k)
         {
           myfile >> value;
-          xsec_map[xsec_names[j]][k].push_back(value);
+          if (myfile.eof())
+            // Check if insufficient number of data values are provided for interp_type=none
+            mooseError(
+                "The number of " + _file_map[_xsec_names[j]] + " values does not match "
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/materials/GenericMoltresMaterial.C:192]: The subject is "The number of" so I believe it's singular.

File: src/materials/NuclearMaterial.C [hunk 2]
```diff
@@ -100,50 +91,42 @@ NuclearMaterial::NuclearMaterial(const InputParameters & parameters)
...
+    _d_chi_p_d_temp[_qp][i] = _xsec_map["CHI_P"][i][0];
+    _d_chi_d_d_temp[_qp][i] = _xsec_map["CHI_D"][i][0];
   }
   for (decltype(_num_groups) i = 0; i < _num_groups * _num_groups; ++i)
   {
-    _gtransfxs[_qp][i] = _xsec_linear_interpolators["GTRANSFXS"][i].sample(dummy_temp);
-    _d_gtransfxs_d_temp[_qp][i] =
-        _xsec_linear_interpolators["GTRANSFXS"][i].sampleDerivative(dummy_temp);
+    _gtransfxs[_qp][i] = _xsec_map["GTRANSFXS"][i][0];
+    _d_gtransfxs_d_temp[_qp][i] = _xsec_map["GTRANSFXS"][i][0];
```
[the remark this answers is not in the document]
Comment (smpark7) [reply src/materials/NuclearMaterial.C:118]: In this case, the user provides group constant data at only one temperature. Taking the derivative with respect to temperature doesn't make sense, but this variable is required by Moltres kernels. I can't think of an elegant way to make `_d_gtransfxs_d_temp` optional, so I'm keeping the existing behavior of passing the same value.

**Comment by smpark7** [comment]
@samgdotson This is ready for merging now

**Review by samgdotson — APPROVED** [review]
lgtm

**Pull request merged**